### PR TITLE
added async version of the Steam and Dota2 clients

### PR DESCRIPTION
**Pull request by OleksiiButeskul**
**Description:** For reactive applications we need to have asynchronous callbacks handling. In such case we can build reactive streams on top of them.

### CODE REVIEW

File: steam-client/src/main/java/com/avenga/steamclient/model/steam/gamecoordinator/dota/match/DotaPicksBans.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.NoArgsConstructor;
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class DotaPicksBans {
+    private boolean isPick;
+    private int heroId;
+    private int team; // 0 - Radiant, 1 - Dire
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/model/steam/gamecoordinator/dota/match/DotaPicksBans.java]: Why regular comment? make that javadoc

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private Date updateTime;
+    private Collection<Message> messages;
+
+    public MarketingMessageCallback(MsgClientMarketingMessageUpdate2 body, byte[] payload) {
+        updateTime = new Date(body.getMarketingMessageUpdateTime() * 1000L);
+
+        List<Message> msgList = new ArrayList<>();
+
+        try (BinaryReader binaryReader = new BinaryReader(new MemoryStream(payload))) {
+            for (int i = 0; i < body.getCount(); i++) {
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: emm. should it really be good-old for loop? forEach mb?

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MarketingMessageCallback.class);
+
+    private Date updateTime;
+    private Collection<Message> messages;
+
+    public MarketingMessageCallback(MsgClientMarketingMessageUpdate2 body, byte[] payload) {
+        updateTime = new Date(body.getMarketingMessageUpdateTime() * 1000L);
+
+        List<Message> msgList = new ArrayList<>();
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: how big this list might be? Mb it is a good ide to pre-allocate bigger list than the default (size 10)

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+     * Represents a single marketing message.
+     */
+    @Getter
+    public static class Message {
+        private GlobalID id;
+        private String url;
+        private EnumSet<EMarketingMessageFlags> flags;
+
+        Message(byte[] data) {
+            try (BinaryReader br = new BinaryReader(new ByteArrayInputStream(data))) {
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: as far as I remember we can now use var inside try with resource

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/UpdateMachineAuthCallback.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        oneTimePassword.setSharedSecret(message.getOtpSharedsecret().toByteArray());
+        oneTimePassword.setTimeDrift(message.getOtpTimedrift());
+    }
+
+    /**
+     * Represents various one-time-password details.
+     */
+    @Getter
+    @Setter
+    public static class OTPDetails {
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/UpdateMachineAuthCallback.java]: emmm? why inner class?

File: steam-client/src/main/java/com/avenga/steamclient/model/steam/gamecoordinator/dota/match/DotaPicksBans.java (hunk 1 above)
Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/model/steam/gamecoordinator/dota/match/DotaPicksBans.java]: Fixed

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java (hunk 2 above)
Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: We know only amount of the messages stored in payload byte array and we don't have list of the elements to iterate.

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java (hunk 3 above)
Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: Added pre-allocated size

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java (hunk 4 above)
Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: Rewrote to vars.

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/UpdateMachineAuthCallback.java (hunk 5 above)
Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/UpdateMachineAuthCallback.java]: Inner classes was extracted to normal. It was copy past from library.

File: steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * This callback is fired when the client receives a marketing message update.
+ */
+@Getter
+public class MarketingMessageCallback extends BaseCallbackMessage {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MarketingMessageCallback.class);
+
+    private Date updateTime;
```
Comment (andyislegend) [inline steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: khm. its 2k20. use of java.util.Date can cause a cancer

Comment (OleksiiButeskul) [reply steam-client/src/main/java/com/avenga/steamclient/steam/asyncclient/steamuser/callback/MarketingMessageCallback.java]: Fixed in all places

**Pull request merged**